### ISSUE TITLE
vmalert: automatically retry recording rules execution on network issues

### ISSUE THREAD

**Issue by dmitryk-dk** [VictoriaMetrics/VictoriaMetrics#4043]
### Is your feature request related to a problem? Please describe

Recording rule evaluation error results in a missing sample in produced time series. Errors might happen due to various reasons: network issues, datasource unavailability. For example, evaluation errors might happen during VM cluster component rolling restarts.

### Describe the solution you'd like

vmalert can automatically re-evaluate recording rules for temporary errors. When the error happens, vmalert could create a queue for rules which need to be re-tried later. Since each evaluation happens at a fixed timestamp, such a queue should consist of a rule and a timestamp at which it needs to be re-evaluated. By default, vmalert can maintain such a queue for 5 evaluations or for 5minutes(maybe this can be defined via flag) before its entries become stale and removed.

### Describe alternatives you've considered

User can use `replay` mode like it described in the [documentation](https://docs.victoriametrics.com/vmalert.html?highlight=datasource.lookback#rules-backfilling)

**Comment by hagen1778** [comment]
> But if some alerting rules are based on the results of recording rules, when we successfully re-evaluate the recording rules afterwards, it could lead to different results of alerting rule which was evaluated before, and user who using raw query to traceback the alert may find the result is wrong.

That's a very good point @Haleygo! I didn't think of this. We do not promise vmalert will respect chaining - see https://docs.victoriametrics.com/vmalert.html#limitations
But I think it would be good at least to match if recording rule name is used within its group by other rules. If it is used, we can't "replay" it afterwards. If this rule is used in other groups - well, we have no power on this and it didn't work correctly anyway. 
To sum up:
1. lets try replay for RR only
2. lets not replay RR which are mentioned in `expr` field of other rules within their group
3. lets add a metric to show the queue size and total number of replayed iterations
4. lets add a group-level param for max tolerable delay when replay is possible

**Comment by XMingG99** [comment]
So when is this problem expected to be solved?

**Comment by hagen1778** [comment]
Hello @XMingG99! It is not in priority list for now. But we're open for contributions.

**Comment by redhatua** [comment]
@dmitryk-dk @tenmozes can this be implemented? It would be nice to have some retry mechanisms.

**Comment by makasim** [comment]
@Haleygo Wouldn't running vmalert in [HA mode](https://docs.victoriametrics.com/victoriametrics/vmalert/#ha-vmalert) address this issue?

**Comment by Haleygo** [comment]
> [@Haleygo](https://github.com/Haleygo) Wouldn't running vmalert in [HA mode](https://docs.victoriametrics.com/victoriametrics/vmalert/#ha-vmalert) address this issue?

No, running HA can only improve reliability when a single vmalert instance fails to connect to the datasource due to a network issue. However, if the datasource itself is unavailable for a period of time (due to an incident or maintenance), or if both vmalert instances cannot connect to the datasource, all evaluations will fail and rule results will have gaps.
To achieve optimal reliability, it is recommended to deploy multiple VictoriaMetrics instances across different zones and expose a global read endpoint as the datasource for vmalert (similar to the approach described in the [victoria-metrics-distributed chart](https://docs.victoriametrics.com/helm/victoria-metrics-distributed/#chart-details)), while also running HA vmalert instances in different zones. This setup, however, requires all components to be highly available, which introduces additional complexity and cost.
Supporting a retry mechanism inside vmalert could be a cost-effective solution for users who do not wish to maintain a large-scale deployment. Nonetheless, reliability still cannot be fully guaranteed, especially in scenarios with frequent failed evaluations. 

Given these considerations, the enhancement is still not in priority list. It's acceptable for TSDB like VictoriaMetrics to have occasional evaluation failures(similar to the scrape behavior in vmagent). If evaluations are consistently failing or users expect to have results during incidents or maintenance, it is advisable to investigate and resolve the root cause (such as network issues or resource shortage) or implement the recommended HA setup for comprehensive reliability.
